engine/apps/schedules/tests/test_on_call_schedule.py

```diff
@@ -2552,3 +2509,165 @@ def test_filter_events_ical_duplicated_uid(make_organization, make_user_for_orga
+    next_shift_start = on_call_shift.start + timezone.timedelta(days=1)
+    upcoming_shift = upcoming_shifts[0]
+    assert upcoming_shift is not None and upcoming_shift["start"] == next_shift_start
+    for shifts in (passed_shifts, current_shifts, upcoming_shifts):
+        for shift in shifts:
+            users = {u["pk"] for u in shift["users"]}
+            assert admin.public_primary_key in users
+
+    passed_shifts, current_shifts, upcoming_shifts = schedule.shifts_for_user(other_user, now)
+    assert len(passed_shifts) == 0
+    assert len(current_shifts) == 0
+    assert len(upcoming_shifts) == 0
+
+
+@pytest.mark.django_db
+def test_shifts_for_user_only_two_users_with_shifts(
+    make_organization,
+    make_user_for_organization,
+    make_schedule,
+    make_on_call_shift,
+):
+    organization = make_organization()
+    current_user = make_user_for_organization(organization)
+    user2 = make_user_for_organization(organization)
+    schedule = make_schedule(organization, schedule_class=OnCallScheduleWeb)
+
+    now = timezone.now()
+    today = now.replace(hour=0, minute=0, second=0, microsecond=0)
+    start_date = today - timezone.timedelta(days=2)
+    days = 7
+
+    data = {
+        "start": now + timezone.timedelta(hours=1),
+        "rotation_start": now + timezone.timedelta(hours=1),
+        "duration": timezone.timedelta(hours=2),
+        "priority_level": 1,
+        "frequency": CustomOnCallShift.FREQUENCY_DAILY,
+        "schedule": schedule,
+    }
+    on_call_shift = make_on_call_shift(
+        organization=organization, shift_type=CustomOnCallShift.TYPE_ROLLING_USERS_EVENT, **data
+    )
+    on_call_shift.add_rolling_users([[current_user]])
+
+    # shift with another user
+    data = {
+        "start": start_date + timezone.timedelta(hours=10),
+        "rotation_start": start_date + timezone.timedelta(hours=10),
+        "duration": timezone.timedelta(hours=24),
+        "priority_level": 1,
+        "frequency": CustomOnCallShift.FREQUENCY_DAILY,
+        "schedule": schedule,
+    }
+    on_call_shift = make_on_call_shift(
+        organization=organization, shift_type=CustomOnCallShift.TYPE_ROLLING_USERS_EVENT, **data
+    )
+    on_call_shift.add_rolling_users([[user2]])
+
+    schedule.refresh_ical_final_schedule()
+
+    passed_shifts, current_shifts, upcoming_shifts = schedule.shifts_for_user(current_user, start_date, days)
+    assert len(passed_shifts) == 0
+    assert len(current_shifts) == 0
+    assert len(upcoming_shifts) == 5
+    for shift in upcoming_shifts:
+        users = {u["pk"] for u in shift["users"]}
+        assert current_user.public_primary_key in users
+
+    passed_shifts, current_shifts, upcoming_shifts = schedule.shifts_for_user(user2, start_date, days)
+    assert len(passed_shifts) > 0
+    assert len(current_shifts) > 0
+    assert len(upcoming_shifts) > 0
+    for shifts in (passed_shifts, current_shifts, upcoming_shifts):
```

Comment:
Besides checking the right user is there, it would be nice to test the past shifts are ended events, current are in progress and upcoming are in the future.